dev/scripts/co-branch.sh

```diff
@@ -0,0 +1,116 @@
+#!/bin/bash
+
+set -e
+
+ME=$(basename "$0")
+
+REMOTE_NAME=${1:-origin}
+BRANCH_NAME=${2:-main}
+OPT_PULL=${3:-0}
+
+usage() {
+  echo "USAGE:"
+  echo "  $ME [REMOTE:origin] [BRANCH:main] [pull]"
+  echo
+  echo "By default $ME will fetch the latest changes for every"
+  echo "submodule and directly checkout the $REMOTE_NAME's $BRANCH_NAME branch."
+  echo "This will leave them in detached HEAD state."
+  echo "Use -p or --pull to instead forward the local $BRANCH_NAME branch."
+}
+
+set_remote() {
+  git ls-remote --exit-code "$REMOTE_NAME" &>/dev/null ||
+    REMOTE_NAME=origin
+}
+
+verify_branch() {
+  git fetch $REMOTE_NAME
+  git rev-parse --verify remotes/"$REMOTE_NAME"/"$BRANCH_NAME" &>/dev/null || BRANCH_NAME=main
+}
+
+pull_latest_commit() {
+  if [ "$OPT_PULL" == "0" ]; then
+    echo "git fetch $REMOTE_NAME && git switch $BRANCH_NAME ..."
+    git fetch "$REMOTE_NAME" &&
+    git checkout "$REMOTE_NAME/$BRANCH_NAME"
+  else
+    echo "git checkout $BRANCH_NAME && git pull --ff-only $REMOTE_NAME $BRANCH_NAME ..."
+    git checkout "$BRANCH_NAME" &&
+    git pull --ff-only "$REMOTE_NAME" "$BRANCH_NAME" || {
+      echo "ERROR: make sure a local branch $BRANCH_NAME exists and can be fast-forwarded to $REMOTE_NAME"
+      exit 1
+    }
+  fi
+}
+
+run_switch () {
+    # First check for changes and stash them if wanted.
+    echo "Check for changes..."
+    if [[ `git status --porcelain --ignore-submodules --untracked-files=no` ]]; then
+        echo "The repository $mod has changes. Stash them? [y/n]"
+        read decision
+        if [ $decision == "y" ]; then
+            git stash
+        else
+            exit 0
+        fi
+    fi
+    for mod in $(git submodule status | awk '{print $2}'); do
```

Comment:
Small note, but couldn't you use `git submodule foreach` instead? Seems simpler

Reply:
For singular commands yes.
But when doing loops with multiple command per submodule and - as done here - getting user input, doing this is often easier to type and read IMO.
Also dealing with escaping and variables is more straightforward.